### ISSUE TITLE
DIY usage question

### ISSUE THREAD

**Issue by mclarsen** [diatomic/diy#68]
I am having difficulty using DIY. Either I am not fully understanding how to use reduce and all-to-all, or there may be a bug. I have been using DIY for my image compositing, and it works in most cases, but in a few configurations, I am getting MPI errors and seg faults.

I have simplified my code and posted it [here](https://github.com/mclarsen/diy_tests)

In the readme there are two test that fail across machines and compilers for me. Can you take a look and see if I am doing something wrong?

**Comment by mrzv** [comment]
The short answer is these are bugs. I need to dig in to see where exactly they come from, but I have a hunch it comes from the fact that `k = 8`, but the block counts have strange prime factors (which decomposer uses when forming a grid of blocks). However, this should still work, so it's a bug, as far as I'm concerned. I'll look into it.

How urgent is it? (The quick short-term fix should be setting `k = 2`.)

One more thing to watch out for. Your `m_int` in `IntBlock` is a reference, `int&`. It works in this particular case, but in general this could be dangerous. I would change it to `int`, just to avoid potential problems down the line.

**Comment by mrzv** [comment]
Yeah, the bug is in `RegularPartners` class. Something is going wrong in group arrangement because of the prime factors.

**Comment by mclarsen** [comment]
Most configurations are working, but I have come across cases in while running a performance study where I encountered this or some other error. 189 was the smallest process count I found where I could duplicated.

This is not super urgent, i.e., I am not up against any deadline at the moment. I figured it would be easy enough to track down for someone in the know.

Is the all_to_all example the same problem or is that something different. There, my stack trace says it died in serialization.

I was passing it by reference because I didn't want to copy the entire int (a stand in for a class with color and depth buffers). In the greater scheme of things, its not a lot of memory, so if you advise the copy, I will head the warning.
